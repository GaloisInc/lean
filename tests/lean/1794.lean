inductive test : nat -> list nat -> Prop
| zero: test 0 list.nil
--n remains even
| nil: forall {n: nat}, test n list.nil -> test (n+2) list.nil
--n flips between even and odd
| cons: forall {n i: nat} {is: list nat}, test n is -> test (n+3) (list.cons i is)

lemma example3 : forall (n m: nat), test (n+n) [m] -> false :=
begin
intros n m,
generalize def_n' : n + n = n',
generalize def_is : [m] = is,
intro h,
revert def_n' def_is,
cases h; try {intros, contradiction}, -- smart unfolding prevents the generation of unwieldy terms,
trace_state,
have : nat.succ (nat.add h_n (nat.add 2 0)) = h_n + 3, from rfl,
simp [this], intro h_1,
have : n + n = h_n + 3 â†’ false, from sorry,
intros,
exact this h_1,
end
