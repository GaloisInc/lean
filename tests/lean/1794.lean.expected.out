case test.cons
n m h_n h_i : ℕ,
h_is : list ℕ,
h_a : test h_n h_is,
h : test (nat.succ (nat.add h_n (nat.add 2 0))) (h_i :: h_is)
⊢ n + n = nat.succ (nat.add h_n (nat.add 2 0)) → [m] = h_i :: h_is → false
1794.lean:8:0: warning: declaration 'example3' uses sorry
